zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -149,6 +159,14 @@ void validateSession(ServerCnxn cnxn, long clientId, int timeout) throws IOExcep
      * @throws IOException
      */
     void writePacket(QuorumPacket pp, boolean flush) throws IOException {
+        if (asyncSending) {
+            sender.queuePacket(pp);
```

Comment:
what about this 'flush' parameter ?

Reply:
If packet are sent asynchronously, flush is no use. If async sending is turned off, then flush is passed on to writePacketNow() and will cause the output buffer flushed.